File: frontend/features/analytics/schemas.ts
import { z } from "zod";

export const periodTypeSchema = z.enum(["day", "week", "month"]);

export const emptyStateSchema = z
  .object({
    code: z.string(),
    message: z.string()
  })
  .nullable();

export const namedTimeTotalSchema = z
  .object({
    label: z.string(),
    total_minutes: z.number().int().nonnegative(),
    display_total: z.string()
  })
  .nullable();

export const dashboardPeriodSchema = z.object({
  period_type: periodTypeSchema,
  anchor_date: z.string(),
  start_date: z.string(),
  end_date: z.string(),
  label: z.string()
});

export const dailyLogListItemSchema = z.object({
  date: z.string(),
  total_minutes: z.number().int().nonnegative(),
  display_total: z.string(),
  task_count: z.number().int().nonnegative(),
  top_category: namedTimeTotalSchema
});

export const periodTimeTimelinePointSchema = z.object({
  date: z.string(),
  total_minutes: z.number().int().nonnegative(),
  display_total: z.string(),
  has_tasks: z.boolean()
});

export const categoryBreakdownItemSchema = z.object({
  label: z.string(),
  total_minutes: z.number().int().nonnegative(),
  display_total: z.string(),
  share_of_total: z.number().min(0).max(1),
  is_other: z.boolean()
});

export const tagBreakdownItemSchema = z.object({
  label: z.string(),
  total_minutes: z.number().int().nonnegative(),
  display_total: z.string(),
  share_label: z.string(),
  is_untagged: z.boolean(),
  is_other: z.boolean()
});

export const dashboardSummarySchema = z.object({
  total_minutes: z.number().int().nonnegative(),
  display_total: z.string(),
  task_count: z.number().int().nonnegative(),
  daily_log_count: z.number().int().nonnegative(),
  highest_time_category: namedTimeTotalSchema,
  average_minutes_per_logged_day: z.number().int().nonnegative().nullable()
});

export const dashboardOverviewSchema = z.object({
  period: dashboardPeriodSchema,
  summary: dashboardSummarySchema,
  daily_logs: z.array(dailyLogListItemSchema),
  period_timeline: z.array(periodTimeTimelinePointSchema),
  category_breakdown: z.array(categoryBreakdownItemSchema),
  tag_breakdown: z.array(tagBreakdownItemSchema),
  tag_total_notice: z.string().nullable(),
  empty_state: emptyStateSchema
});

export const taskTimelineItemSchema = z.object({
  id: z.string().uuid(),
  title: z.string(),
  time_spent_minutes: z.number().int().positive(),
  display_time: z.string(),
  category: z.string(),
  tags: z.array(z.string()),
  note: z.string().nullable().optional(),
  source: z.string(),
  timeline_position: z.number().int().nonnegative()
});

export const dayDetailSchema = z.object({
  date: z.string(),
  total_minutes: z.number().int().nonnegative(),
  display_total: z.string(),
  task_count: z.number().int().nonnegative(),
  tasks: z.array(taskTimelineItemSchema),
  empty_state: emptyStateSchema
});

File: backend/app/modules/analytics/schemas.py
from datetime import date
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field


PeriodType = Literal["day", "week", "month"]


class EmptyState(BaseModel):
    code: str
    message: str


class DashboardPeriod(BaseModel):
    period_type: PeriodType
    anchor_date: date
    start_date: date
    end_date: date
    label: str


class NamedTimeTotal(BaseModel):
    label: str
    total_minutes: int = Field(ge=0)
    display_total: str


class DashboardSummary(BaseModel):
    total_minutes: int = Field(ge=0)
    display_total: str
    task_count: int = Field(ge=0)
    daily_log_count: int = Field(ge=0)
    highest_time_category: NamedTimeTotal | None = None
    average_minutes_per_logged_day: int | None = Field(default=None, ge=0)


class DailyLogListItem(BaseModel):
    date: date
    total_minutes: int = Field(ge=0)
    display_total: str
    task_count: int = Field(ge=0)
    top_category: NamedTimeTotal | None = None


class PeriodTimeTimelinePoint(BaseModel):
    date: date
    total_minutes: int = Field(ge=0)
    display_total: str
    has_tasks: bool


class CategoryBreakdownItem(BaseModel):
    label: str
    total_minutes: int = Field(ge=0)
    display_total: str
    share_of_total: float = Field(ge=0, le=1)
    is_other: bool


class TagBreakdownItem(BaseModel):
    label: str
    total_minutes: int = Field(ge=0)
    display_total: str
    share_label: str
    is_untagged: bool
    is_other: bool


class DashboardOverview(BaseModel):
    period: DashboardPeriod
    summary: DashboardSummary
    daily_logs: list[DailyLogListItem] = Field(default_factory=list)
    period_timeline: list[PeriodTimeTimelinePoint] = Field(default_factory=list)
    category_breakdown: list[CategoryBreakdownItem] = Field(default_factory=list)
    tag_breakdown: list[TagBreakdownItem] = Field(default_factory=list)
    tag_total_notice: str | None = None
    empty_state: EmptyState | None = None


class DashboardTaskRow(BaseModel):
    id: UUID
    log_date: date
    title: str
    time_spent_minutes: int = Field(gt=0)
    category: str
    tags: list[str] = Field(default_factory=list)
    source: str
